jni/src/Foreign/JNI/Types.hs

```diff
@@ -148,22 +152,31 @@ instance SingI 'Void where
 type a <> g = 'Generic a g
 
 -- | Type indexed Java Objects.
-newtype J (a :: JType) = J (Ptr (J a))
-  deriving (Eq, Show, Storable)
+newtype J (a :: JType) = J (ForeignPtr (J a))
+  deriving (Eq, Show)
 
 type role J representational
 
+-- | Creates a J value from a Java reference.
+newJ :: Ptr (J ty) -> IO (J ty)
+newJ = fmap J . newForeignPtr_
+
+-- | Provides access to the underlying reference and makes sure no finalizer
+-- will be called before the given action returns.
+withJ :: J ty -> (Ptr (J ty) -> IO a) -> IO a
+withJ (J fp) = withForeignPtr fp
```

Comment:
"J" is never mentioned at the value level. It's just a pun to keep types non redundant. Say `withObjectPtr`.